pennylane/operation.py

```diff
@@ -383,6 +383,105 @@ def circuit(angle):
+                return qml.math.array([[c, -1j * s], [-1j * s, c]])
+
+        If we passed a broadcasted argument ``theta`` of shape ``(batch_size,)`` to this method,
+        which would have one instead of zero dimensions, ``cos`` and ``sin`` would correctly
+        be applied elementwise.
+        We would also obtain the correct matrix with shape ``(2, 2, batch_size)``.
+        However, the broadcasting axis needs to be the *first* axis by convention, so that we need
+        to move the broadcasting axis--if it exists--to the front before returning the matrix:
+
+        .. code-block:: python
+
+            @staticmethod
+            def compute_matrix(theta):
+                '''Broadcasting axis ends up in the correct leading position.'''
+                c = qml.math.cos(theta / 2)
+                s = qml.math.sin(theta / 2)
+                mat = qml.math.array([[c, -1j * s], [-1j * s, c]])
+                # Check whether the input has a broadcasting axis
+                if qml.math.ndim(theta)==1:
+                    # Move the broadcasting axis to the first position
+                    return qml.math.moveaxis(mat, 2, 0)
+                return mat
+
+        Adapting ``compute_eigvals`` to broadcasting looks similar.
+
+        Usually no major changes are required for ``compute_decomposition``, but we need
+        to take care of the correct mapping of input arguments to the operators in the
+        decomposition. As an example, consider the operator that represents a layer of
+        ``RX`` rotations with individual angles for each rotation. Without broadcasting,
+        it takes one onedimensional array, i.e. ``ndim_params=(1,)``.
+        Its decomposition, which is a convenient way to support this custom operation
+        on all devices that implement ``RX``, might look like this:
+
+        .. code-block:: python
+
+            @staticmethod
+            def compute_decomposition(theta, wires):
+                '''Iterate over the first axis of theta.'''
+                decomp_ops = [qml.RX(x, wires=w) for x, w in zip(theta, wires)]
+                return decomp_ops
+
+        If ``theta`` is a broadcasted argument, its first axis is the broadcasting
+        axis and we would like to iterate over the *second* axis within the ``for``
+        loop instead. This is easily achieved by adding a transposition of ``theta``
+        that switches the axes in this case. Conveniently this does not have any
+        effect in the non-broadcasted case, so that we do not need to handle two
+        cases separately.
+
+        .. code-block:: python
+
+            @staticmethod
+            def compute_decomposition(theta, wires):
+                '''Iterate over the last axis of theta, which is also the first axis
+                or the second axis without and with broadcasting, respectively.'''
+                decomp_ops = [qml.RX(x, wires=w) for x, w in zip(qml.math.T(theta), wires)]
+                return decomp_ops
```

Comment:
Worth adding documentation on how `_check_batching` is supposed to work, answering such questions as:
* when does `_check_batching` need to be called? (any time the data is modified)
* what attributes does it set?
* What do those attributes look like if `_check_batching` is not called?
* What do those attributes look like if the Operator does not support batching?
* What does `_check_batching` do if the Operator does not support batching?

Reply:
Great idea, thanks!

Reply:
I added a paragraph on `_check_batching`. Happy to hear your thoughts on it!